modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+		pub(crate) fn on_message_sent_to_bridge(message_size: u32) {
+			// if outbound queue is not overloaded, do nothing
+			if !T::ToBridgeHubSender::is_overloaded() {
+				return
+			}
+
+			// ok - we need to increase the fee factor, let's do that
+			let message_size_factor = FixedU128::from_u32(message_size.saturating_div(1024))
+				.saturating_mul(MESSAGE_SIZE_FEE_BASE);
+			let total_factor = EXPONENTIAL_FEE_BASE.saturating_add(message_size_factor);
+			DeliveryFeeFactor::<T, I>::mutate(|f| {
+				*f = f.saturating_mul(total_factor);
+				*f
+			});
+		}
+	}
+}
+
+/// We'll be using `SovereignPaidRemoteExporter` to send remote messages over the sibling/child
+/// bridge hub.
+type ViaBridgeHubExporter<T, I> = SovereignPaidRemoteExporter<
+	Pallet<T, I>,
+	<T as Config<I>>::ToBridgeHubSender,
+	<T as Config<I>>::UniversalLocation,
+>;
+
+// This pallet acts as the `ExporterFor` for the `SovereignPaidRemoteExporter` to compute
+// message fee using fee factor.
+impl<T: Config<I>, I: 'static> ExporterFor for Pallet<T, I> {
+	fn exporter_for(
+		network: &NetworkId,
+		_remote_location: &InteriorMultiLocation,
+		message: &Xcm<()>,
+	) -> Option<(MultiLocation, Option<MultiAsset>)> {
+		// ensure that the message is sent to the expected bridged network
+		if *network != T::BridgedNetworkId::get() {
+			return None
+		}
+
+		// compute fee amount
+		let mesage_size = message.encoded_size();
+		let message_fee = (mesage_size as u128).saturating_mul(T::ByteFee::get());
+		let fee_sum = T::BaseFee::get().saturating_add(message_fee);
+		let fee = Self::delivery_fee_factor().saturating_mul_int(fee_sum);
+
+		Some((T::SiblingBridgeHubLocation::get(), Some((T::FeeAsset::get(), fee).into())))
+	}
+}
+
+// This pallet acts as the `SendXcm` to the sibling/child bridge hub instead of regular
+// XCMP/DMP transport. This allows injecting dynamic message fees into XCM programs that
+// are going to the bridged network.
+impl<T: Config<I>, I: 'static> SendXcm for Pallet<T, I> {
+	type Ticket = (u32, <T::ToBridgeHubSender as SendXcm>::Ticket);
+
+	fn validate(
+		dest: &mut Option<MultiLocation>,
+		xcm: &mut Option<Xcm<()>>,
+	) -> SendResult<Self::Ticket> {
+		// we won't have an access to `dest` and `xcm` in the `delvier` method, so precompute
+		// everything required here
+		let message_size = xcm
+			.as_ref()
+			.map(|xcm| xcm.encoded_size() as _)
+			.ok_or(SendError::MissingArgument)?;
+
+		// TODO: we can reject large messages here (now they're just dropped at the bridge hub)!!!
+
+		// just use exporter to validate destination and insert instructions to pay message fee
+		// at the sibling/child bridge hub
+		//
+		// the cost will include both cost of: (1) to-sibling bridg hub delivery (returned by
+		// the `Config::ToBridgeHubSender`) and (2) to-bridged bridge hub delivery (returned by
+		// `Self::exporter_for`)
+		ViaBridgeHubExporter::<T, I>::validate(dest, xcm)
```

Comment:
Another dependency we'll need for that is https://github.com/paritytech/cumulus/pull/2342. So idea is:
1) bridge backpressure mechanism involves many chains, so it isn't instant - it may take some to suspend the queue between source AH and source BH. And we can't allow senders to send messages over the bridge for a constant cost while this mechanism activates;
2) so we need to charge dynamic fees for sending messages from source AH and source BH even when we don't have any info about other bridge queues. That's where the mentioned PR should help us - the total transport cost that the sender pays is the `BridgeFee + HrmpFee`. So `BridgeFee` will grow when we have a report about overloaded bridge queues and `HrmpFee` shall be maintained independently